sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,333 @@
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.vendor.grpc.v1p43p2.com.google.protobuf.ByteString;
+import org.apache.beam.vendor.grpc.v1p43p2.io.grpc.stub.StreamObserver;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An outbound data buffering aggregator with size-based buffer and time-based buffer if
+ * corresponding options are set.
+ *
+ * <p>The default size-based buffer threshold can be overridden by specifying the experiment {@code
+ * data_buffer_size_limit=<bytes>}
+ *
+ * <p>The default time-based buffer threshold can be overridden by specifying the experiment {@code
+ * data_buffer_time_limit_ms=<milliseconds>}
+ */
+@SuppressWarnings({
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class BeamFnDataOutboundAggregator {
+
+  public static final String DATA_BUFFER_SIZE_LIMIT = "data_buffer_size_limit=";
+  public static final int DEFAULT_BUFFER_LIMIT_BYTES = 1_000_000;
+  public static final String DATA_BUFFER_TIME_LIMIT_MS = "data_buffer_time_limit_ms=";
+  public static final long DEFAULT_BUFFER_LIMIT_TIME_MS = -1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(BeamFnDataOutboundAggregator.class);
+  private final int sizeLimit;
+  private final long timeLimit;
+  private final Supplier<String> processBundleRequestIdSupplier;
+  private final Map<String, Coder<?>> outputDataCoders;
+  private final Map<KV<String, String>, Coder<?>> outputTimersCoders;
+  private final Map<String, LongAdder> perEndpointDataByteCount;
+  private final Map<KV<String, String>, LongAdder> perEndpointTimersByteCount;
+  private final Map<LogicalEndpoint, Long> perEndpointElementCount;
+  private final StreamObserver<Elements> outboundObserver;
+  private final Map<String, ByteString.Output> dataBuffer;
+  private final Map<KV<String, String>, ByteString.Output> timersBuffer;
+  @VisibleForTesting ScheduledFuture<?> flushFuture;
+  private Long totalByteCounter = 0L;
+  private final Object flushLock = new Object();
+
+  public BeamFnDataOutboundAggregator(
+      PipelineOptions options,
+      Supplier<String> processBundleRequestIdSupplier,
+      StreamObserver<BeamFnApi.Elements> outboundObserver) {
+    this.sizeLimit = getSizeLimit(options);
+    this.timeLimit = getTimeLimit(options);
+    this.outputDataCoders = new HashMap<>();
+    this.outputTimersCoders = new HashMap<>();
+    this.perEndpointDataByteCount = new HashMap<>();
+    this.perEndpointTimersByteCount = new HashMap<>();
+    this.perEndpointElementCount = new HashMap<>();
+    this.outboundObserver = outboundObserver;
+    this.dataBuffer = new HashMap<>();
+    this.timersBuffer = new HashMap<>();
+    if (timeLimit > 0) {
+      this.flushFuture =
```

Comment:
You should add a start method that starts the `flushFuture` (if necessary) once all the registration has finished. This will ensure that the maps are effectively final.

Reply:
Done.

Reply:
Done.